### PR TITLE
DL full event reconstruction

### PR DESCRIPTION
**Pull request by TjarkMiener**
This PR includes three major features:

1. **Energy and arrival direction regression** (Impact parameter and shower maximum opitonal)

    - The regression tasks have been added to CTLearn. Full event reconstruction can be performed in both modes (Singletask Learning and Multitask Learning). However, in the MTL approach the protons are not excluded in the loss calculation and the task weights are fix constant from the config file. This will be solved in a follow-up PR.

2. **Residual blocks**

    - Residual blocks and a stack function have been added to the engine (Basic). This allows the user to set up a custom ResNet architecture from the config file.
    References:
        - [Deep Residual Learning for Image Recognition](https://arxiv.org/pdf/1512.03385.pdf)

3. **Attention mechanisms** 

    - Three different attention mechanisms have been added to Basic, which can be used in all four CTLearn models.
    References:
        - [Squeeze and Excitation Networks](https://arxiv.org/abs/1709.01507)
        - [Concurrent Spatial and Channel Squeeze & Excitation in Fully Convolutional Networks](https://arxiv.org/abs/1803.02579)

Some minor features:
- CTLearn can be run in 'trainandpredict' to predict after training.
- Multiple runs with the same model, but different random seeds can be performed sequentially with one command.
- Minor restructuring of the run_model.py. Separating the data loading (setting up the DL1DH: inputfn() etc.) and the output handling from the main script.
------------------------
This PR requires [DL1DH PR 85](https://github.com/cta-observatory/dl1-data-handler/pull/85).

### CODE REVIEW

**Comment by TjarkMiener** [comment]
This PR would be ready for code review! Thank you, @nietootein @aribrill!

**Comment by TjarkMiener** [comment]
Thanks @LucaRomanato for reporting a bug occur in the subfolder structure. `6733ed0` should fix this!

**Comment by aribrill** [comment]
I am in the process of testing the PR on shevek and will have more comments soon. One issue I came across so far is the installation. I followed the installation instructions in the README. Are they still up to date, e.g. regarding installing DL1DH with pip? I had to run the following commands after setup to get things working:

```bash
source activate ctlearn
conda install -c cta-observatory ctapipe-extra
pip uninstall numpy
pip install numpy
```

File: README.md [hunk 1]
```diff
@@ -120,12 +121,14 @@ Run CTLearn from the command line:
 CTLEARN_DIR=</installation/path>/ctlearn/ctlearn
 python $CTLEARN_DIR/run_model.py myconfig.yml [--mode <MODE>] [--debug] [--log_to_file]
 ```
-`--mode <MODE>`: Set run mode with `<MODE>` as `train`, `predict`, or `load_only`. If not set, defaults to `train`.
+`--mode <MODE>`: Set run mode with `<MODE>` as `train`, `predict`, `trainandpredict`, or `load_only`. If not set, defaults to `train`.
```
Comment (aribrill) [inline README.md:124]:
```suggestion
`--mode <MODE>`: Set run mode with `<MODE>` as `train`, `predict`, `train_and_predict`, or `load_only`. If not set, defaults to `train`.
```

File: ctlearn/default_models/basic.py [hunk 2]
```diff
@@ -11,26 +12,30 @@ def conv_block(inputs, training, params, reuse=None):
                 params['basic']['conv_block']['layers']]
         kernel_sizes = [layer['kernel_size'] for layer in
                 params['basic']['conv_block']['layers']]
+        numbers_list = [layer.get('number', 1) for layer in
+                        params['basic']['conv_block']['layers']]
         max_pool = params['basic']['conv_block']['max_pool']
         bottleneck_filters = params['basic']['conv_block']['bottleneck']
         batchnorm = params['basic']['conv_block'].get('batchnorm', False)
-        
+        attention = params.get('attention', None)
```
Comment (aribrill) [inline ctlearn/default_models/basic.py:20]:
```suggestion
        attention = params.get('attention')
```

File: ctlearn/output_handler.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        # Shower maximum regression
+        if 'showermaximum' in prediction:
+            if i == 0:
+                data['reco_x_max'] = []
+            data['reco_x_max'].append(prediction['showermaximum'][0])
+
+    if prediction_label in list(pd.HDFStore(h5file).keys()):
+        pd.HDFStore(h5file).remove(prediction_label)
+    pd.DataFrame(data=data).to_hdf(h5file, key=prediction_label, mode='a')
+    return
```
Comment (aribrill) [inline ctlearn/output_handler.py:68]: Should something be returned here? Otherwise, there's no need to explicitly return

File: config/example_config.yml [hunk 4]
```diff
@@ -251,26 +252,33 @@ Model:
...
     # in order of class label (0 to n-1).
-    # For a regression label, the value is null.
     # Optional dictionary with string keys and values either null or
     # a list of strings. Default: {}
-    label_names:
-        class_label:
-            - 'proton'
-            - 'gamma'
+    # These options have to be in consonance with the 'event_info' and 
+    # 'transforms' in 'Data'.
```
Comment (aribrill) [inline config/example_config.yml:275]: It would helpful to document what the necessary transforms and event_info are for each task.

File: ctlearn/run_model.py [hunk 5]
```diff
@@ -506,10 +358,66 @@ def write(prediction):
         '--log_to_file',
         action='store_true',
         help="log to a file in model directory instead of terminal")
+    parser.add_argument(
```
Comment (aribrill) [inline ctlearn/run_model.py:361]: Following an offline discussion with @TjarkMiener and @LucaRomanato, the `multiple_runs` command line argument should be removed entirely. It doesn't work well in practice, and the same functionality can be obtained if desired using `run_multiple_configurations.py`. Can you take it out (and update the Readme to match)?

File: ctlearn/run_model.py [hunk 6]
```diff
@@ -506,10 +358,66 @@ def write(prediction):
...
+            run_model(config, mode='train', debug=args.debug, log_to_file=args.log_to_file)
+            for key in config['Prediction']['prediction_file_lists']:
+                with open(args.config_file, 'r') as config_file:
+                    config = yaml.safe_load(config_file)
+                config['Data']['seed'] = random_seed
+                if args.multiple_runs != 1 or args.random_seed != 0:
+                    config['Logging']['add_seed'] = True
+                config['Data']['shuffle'] = False
+                config['Prediction']['prediction_label'] = key
+                run_model(config, mode='predict', debug=args.debug, log_to_file=args.log_to_file)
```
Comment (aribrill) [inline ctlearn/run_model.py:423]:
```suggestion
        if args.mode in ['train', 'train_and_predict', 'load_only']:
            run_model(config, mode=args.mode, debug=args.debug, log_to_file=args.log_to_file)
        if args.mode in ['predict', 'train_and_predict']:
            for key in config['Prediction']['prediction_file_lists']:
                with open(args.config_file, 'r') as config_file:
                    config = yaml.safe_load(config_file)
                config['Data']['seed'] = random_seed
                if args.multiple_runs != 1 or args.random_seed != 0:
                    config['Logging']['add_seed'] = True
                config['Data']['shuffle'] = False
                config['Prediction']['prediction_label'] = key
                run_model(config, mode=args.mode, debug=args.debug, log_to_file=args.log_to_file)
```

File: ctlearn/utils.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
+import importlib
+import logging
+import numpy as np
+import os
+import pkg_resources
+import sys
+import pandas as pd
+import time
+import yaml
```
Comment (aribrill) [inline ctlearn/utils.py:9]:
```suggestion
import importlib
import logging
import os
import pkg_resources
import sys
import time

import numpy as np
import pandas as pd
import yaml
```

File: ctlearn/data_loader.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+import logging
+import numpy as np
+import os
+import pkg_resources
+import sys
+import pandas as pd
+import time
+import yaml
+
+import tensorflow as tf
```
Comment (aribrill) [inline ctlearn/data_loader.py:11]:
```suggestion
import importlib
import logging
import os
import pkg_resources
import sys
import time

import numpy as np
import pandas as pd
import tensorflow as tf
import yaml
```

**Review by aribrill — CHANGES_REQUESTED** [review]
The PR is very nice! I am currently replicating one of the new benchmarks as a test.

Some other comments:
- Please edit the Readme to bring it up to date with the PR, e.g. update the installation instructions with all required dependencies and specified DL1DH version; remove references to deprecated features like multi-stereo mode; and make sure that the command line options listed in the "Run a Model" section are the right ones.
- In utils.py, compute_class_weights() and log_examples() aren't compatible with the new multitask config format. I can commit fixes tomorrow after the benchmark test is done.
- Computing the number of examples when a label is a sequence such as (alt, az) also required a change to DL1DH, which I have requested as cta-observatory/dl1-data-handler#99.

File: ctlearn/run_model.py (hunk 6 above)
Comment (TjarkMiener) [reply ctlearn/run_model.py]: Thanks for the suggestion! I sightly modified it to make it work. The run_model function not have a 'train_and_predict' mode, so we have to call it recursively and manually.

I search for "train" and "predict" in the args.mode string. So it will also work accordingly if the user set the mode to "training", "predicting", "prediction", "train_and_predict", "training_and_predicting" ...
The mode "load_only" is separated.

What do you think?

Comment (aribrill) [reply ctlearn/run_model.py]: Clever! Seems good to me.

File: ctlearn/run_model.py (hunk 5 above)
Comment (TjarkMiener) [reply ctlearn/run_model.py:362]: Done

File: config/example_config.yml (hunk 4 above)
Comment (TjarkMiener) [reply config/example_config.yml:276]: Done

**Comment by TjarkMiener** [comment]
Thanks for the comments @aribrill! I resolved most of them. Regarding the installation, can you specify the ctapipe and dl1dh version you used please? ctapipe-extra should not be required though.

Also thanks for fixing the number of examples of DL1DH. I will have a look into it now.

**Comment by aribrill** [comment]
I used DL1DH v0.8.2, which requires ctapipe 0.8.0.

**Comment by TjarkMiener** [comment]
The correct versions of DL1DH and ctapipe should be installed automatically by conda, when the environment is created. There is no need to git clone the DL1DH repo. The latest commits should fix the installation and I updated the README. Can you please check, if the correct versions are installed on `shevek`? I also added a commit to make ctlearn a command line tool to avoid pointing to the run_model script. This should work: `$ ctlearn myconfig.yml`

**Review by aribrill — CHANGES_REQUESTED** [review]
Looking great! I can see only one small additional change to make resulting from the recent revisions.

Thanks for updating the installation method, I can confirm that it works on shevek (although the compatibility PR is not included). I really like the ability to use ctlearn as a command line tool.

**Comment by aribrill** [comment]
Actually, I have a question. I replicated one of the benchmark models (arr_lowcut_TRN) and would like to plot the predictions for diffuse gammas. I have a predictions file in h5 format. I see that none of the plotting scripts have been updated to be compatible with this new format, however. Are there new scripts that replicate the old functionality or at least some instructions for what to do with this file?

I also see that ctaplot has been added as a dependency, but I can't find where it's used. Is that related in some way?

File: ctlearn/run_model.py [hunk 9]
```diff
@@ -2,103 +2,44 @@
...
+    root_model_dir = model_dir = config['Logging']['model_directory']
     # Create model directory if it doesn't exist already
+    if not os.path.exists(root_model_dir):
+        if mode == 'predict':
+            raise ValueError("Invalid model directory '{}'. "
+            "Must be a path to an existing directory in the predict mode.".format(config['Logging']['model_directory']))
+        os.makedirs(root_model_dir)
+
+    random_seed = config['Data']['seed']
+    model_dir += "/seed_{}".format(random_seed)
```
[the remark this answers is not in the document]
Comment (TjarkMiener) [reply ctlearn/run_model.py]: Done!

**Comment by TjarkMiener** [comment]
Thanks for the fixes! Everything worked fine for the compatibility PR. I uploaded the visualisation notebook. Anything missing?

**Review by aribrill — APPROVED** [review]
Looks good! I updated the particle classification plotting scripts for compatibility with the new predictions output format.

**Comment by TjarkMiener** [comment]
Thanks for updating the plotting scripts!
